bip-unknown-var-budget-script.mediawiki

```diff
@@ -0,0 +1,722 @@
+# Pop A off the stack.
+# Remove BEGIN bytes from the front of A (all bytes if BEGIN greater than length of A).
+# If length(A) is greater than value(LEN), truncate A to length value(LEN).
+# Push A onto the stack.
+|-
+|OP_LEFT
+|128
+|2
+|
+# Pop OFFSET off the stack.
+# Pop A off the stack.
+# If length(A) is greater than value(OFFSET), truncate A to length value(OFFSET).
+# Push A onto the stack.
+|-
+|OP_RIGHT
+|129
+|2
+|
+# Pop OFFSET off the stack.
+# Pop A off the stack.
+# Remove value(OFFSET) bytes from the front of A, or all bytes if value(OFFSET) is greater than the length(A).
+# Push A onto the stack.
+|}
+
+Bit operation opcodes:
+
+{|
+! Opcode
+! Value
+! Required Stack Elements
+! Definition
+|-
+|OP_INVERT
+|131
+|1
+|
+# Pop A off the stack.
+# For each byte in A, replace it with that byte bitwise XOR 0xFF (i.e. invert the bits)
+# Push A onto the stack.
+|-
+|OP_AND
+|132
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in A (the longer operand): bitwise AND it with the equivalent byte in B (or 0 if past end of B)
+# Push A onto the stack.
+|-
+|OP_OR
+|133
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in B (the shorter operand): bitwise OR it with the equivalent byte in A.
+# Push A onto the stack.
+|-
+|OP_XOR
+|134
+|2
+|
+# Pop B off the stack.
+# Pop A off the stack.
+# If B is longer than A, swap B and A.
+# For each byte in B (the shorter operand): exclusive OR it with the equivalent byte in A.
+# Push A onto the stack.
+|}
+
+Bitshift opcodes:
+
+Note that these are raw bitshifts, unlike the sign-preserving arithmetic shifts in Bitcoin v0.3.0, and as such they also do not truncated trailing zeroes from results: they are renamed OP_UPSHIFT (nee OP_LSHIFT) and OP_DOWNSHIFT (nee OP_RSHIFT).
+
+Note also that OP_UPSHIFT can produce huge results, and so must be checked for limits prior to evaluation.  It is also carefully defined to avoid reallocating twice (reallocating to prepend bytes, then again to append a single byte) which has the practical advantage of being about to share the same downward bitshift routine as OP_DOWNSHIFT.
+
+{|
+! Opcode
+! Value
+! Required Stack Elements
+! Definition
+|-
+|OP_UPSHIFT
+|152
+|2
+|
+# Pop BITS off the stack.
+# Pop A off the stack.
+# If A shifted by value(BITS) would exceed the individual stack limit, fail.
+# If value(BITS) % 8 == 0: simply prepend value(BITS) / 8 zeroes to A.
+# Otherwise: prepend (value(BITS) / 8) + 1 zeroes to A, then shift A *down* (8 - (value(BITS) % 8)) bits.
+# Push A onto the stack.
+|-
+|OP_DOWNSHIFT
+|153
+|2
+|
+# Pop BITS off the stack.
+# Pop A off the stack.
+# For OFF from value(BITS) / 8 to length(A)-1:
+    # Copy each byte from OFF, shifting by (BITS mod 8), to BITS / 8 bytes earlier in A.
+# Truncate A to remove value(OFF) bytes (or all, if value(OFF) > length(A)).
+# Push A onto the stack.
+|}
```

Comment:
I cannot parse this description. OFF appears to be an index. Are you supposed to copy A[OFF] instead of OFF into A? Also would be helpful to have concrete index instead of "earlier".